File: codes/naeun/1277_countsquaresubmatriceswithallones_1.java
class Solution {
    public int countSquares(int[][] matrix) {
    // 확장된 dp 사용(out of index 처리)
    int dp[][] = new int[matrix.length+1][matrix[0].length+1];
    int ans = 0; 
    for(int i=0; i<matrix.length; i++){
      for(int j=0; j<matrix[0].length; j++){
        // matrix i,j는 dp의 i+1, j+1과 같음.
        if(matrix[i][j]==1){
            dp[i+1][j+1] = 1 + Math.min(dp[i][j], Math.min(dp[i+1][j], dp[i][j+1]));
        }
        ans += dp[i+1][j+1];
      }  
    }  
    return ans;
    }
}

File: codes/jiheon/1277_countsquaresubmatriceswithallones_1.rs
impl Solution {
    pub fn count_squares(matrix: Vec<Vec<i32>>) -> i32 {
        let (n, m) = (matrix.len(), matrix[0].len());
        let mut ans = 0;
        let mut DP = vec![0; m];
        let mut prev = 0;

        for i in 0..n {
            for j in 0..m {
                let tmp = DP[j];

                if matrix[i][j] == 1 {
                    if i == 0 || j == 0 {
                        DP[j] = 1;
                    } else {
                        DP[j] = DP[j].min(DP[j - 1]).min(prev) + 1;
                    }
                    ans += DP[j];
                } else {
                    DP[j] = 0;
                }
                prev = tmp;
            }
        }

        ans
    }
}